llpc/util/llpcElfWriter.cpp

```diff
@@ -735,6 +737,286 @@ template <class Elf> void ElfWriter<Elf>::updateMetaNote(Context *pContext, cons
   pNewNote->data = data;
 }
 
+// =====================================================================================================================
+// Get the count of relocs in the reloc section
+//
+// @param relocSection : Reloc section
+template <class Elf> size_t ElfWriter<Elf>::numRelocs(const SectionBuffer *relocSection) {
+  size_t relocCount = 0;
+  if (relocSection->secHead.sh_size > 0) {
+    relocCount = static_cast<unsigned>(relocSection->secHead.sh_size / relocSection->secHead.sh_entsize);
+  }
+
+  return relocCount;
+}
+
+// =====================================================================================================================
+// Get the start pos of relocs used by PS in the reloc section
+//
+// @param relocSection : Reloc section
+// @param psIsaOffset : Ps ISA offset in the elf bin
+template <class Elf> size_t ElfWriter<Elf>::getRelocPsStartPos(const SectionBuffer *relocSection, size_t psIsaOffset) {
+
+  size_t relocCount = numRelocs(relocSection);
+  if (relocCount == 0)
+    return 0;
+
+  size_t psStartPos = 0;
+  auto relocs = reinterpret_cast<typename Elf::Reloc *>(const_cast<uint8_t *>(relocSection->data));
+  for (; psStartPos < relocCount; ++psStartPos)
+    if (relocs[psStartPos].r_offset >= psIsaOffset)
+      break;
+
+  return psStartPos;
+}
+
+// =====================================================================================================================
+// Create a new empty section based on the input section
+//
+// @param name : Name of the new section
+// @param inputSection : The input section
+// @param [out] newSection : New section returned
+template <class Elf>
+void ElfWriter<Elf>::createNewSection(const char *sectionName, SectionBuffer *inputSection, SectionBuffer *newSection) {
```

Comment:
I have tried return `SectionBuffer *` but got the following error,
```
/home/llpc/llpc/util/llpcElfWriter.cpp:807:1: error: ‘SectionBuffer’ does not name a type
  807 | SectionBuffer* ElfWriter<Elf>::createNewSection(const char *sectionName, SectionBuffer *inputSection) {
      | ^~~~~~~~~~~~~
```
It seems that we need to use `ElfSectionBuffer<typename Elf::SectionHeader> *` instead, can you accept changing the function to `ElfSectionBuffer<typename Elf::SectionHeader> * createNewSection(const char *sectionName, SectionBuffer *inputSection)` ?

Reply:
If you make the `SectionBuffer` typedef public, I think you can just do:
```
// in .h
SectionBuffer createNewSection(const char *sectionName, SectionBuffer *inputSection);
// in .cpp
ElfWriter<Elf>::SectionBuffer ElfWriter<Elf>::createNewSection(const char *sectionName, SectionBuffer *inputSection) {
```

Note that I propose to return the new section, not a pointer. Since the `SectionBuffer` is not address-taken from what I can tell, I don't see why returning a pointer would be preferred.

Reply:
Okay, I will try this.